src/consistent_request.rs

```diff
@@ -0,0 +1,222 @@
+    execution_mode: Arc<Mutex<ExecutionMode>>,
+}
+
+impl ConsistentRequest {
+    /// Creates a new `ConsistentRequest` instance.
+    pub fn new(
+        method: String,
+        l2_client: HttpClient,
+        builder_client: HttpClient,
+        probes: Arc<Probes>,
+        execution_mode: Arc<Mutex<ExecutionMode>>,
+    ) -> Self {
+        let (req_tx, mut req_rx) = watch::channel(None);
+        let (res_tx, mut res_rx) = watch::channel(None);
+        req_rx.mark_unchanged();
+        res_rx.mark_unchanged();
+
+        let has_disabled_execution_mode = Arc::new(AtomicBool::new(false));
+
+        let manager = Self {
+            method,
+            l2_client,
+            builder_client,
+            has_disabled_execution_mode,
+            req_tx,
+            res_rx,
+            probes,
+            execution_mode,
+        };
+
+        let clone = manager.clone();
+        tokio::spawn(async move {
+            let mut attempt: Option<JoinHandle<_>> = None;
+            loop {
+                req_rx
+                    .changed()
+                    .await
+                    .expect("channel should always be open");
+
+                if let Some(attempt) = attempt {
+                    if !attempt.is_finished() {
+                        error!(target: "proxy::call", method = clone.method, message = "request cancelled");
+                        attempt.abort();
+                    }
+                }
+
+                let (parts, body) = req_rx
+                    .borrow_and_update()
+                    .as_ref()
+                    .expect("value should always be Some")
+                    .clone();
+
+                let mut clone = clone.clone();
+                let res_tx_clone = res_tx.clone();
+                attempt = Some(tokio::spawn(async move {
+                    clone
+                        .send_with_retry_cancel_safe(parts, body, res_tx_clone)
+                        .await
+                }));
+            }
+        });
+
+        manager
+    }
+
+    /// This function may be cancelled at any time from an incoming request.
+    /// We should ensure we're in a valid state at all await points.
+    async fn send_with_retry_cancel_safe(
+        &mut self,
+        parts: request::Parts,
+        body: Vec<u8>,
+        res_tx: watch::Sender<Option<Result<(response::Parts, Vec<u8>), BoxError>>>,
+    ) -> eyre::Result<()> {
+        // We send the l2 request first, because we need to avoid the situation where the
+        // l2 fails and the builder succeeds. If this were to happen, it would be too dangerous to
+        // return the l2 error response back to the caller, since the builder would now have an
+        // invalid state. We can return early if the l2 request fails.
+        let mut manager = self.clone();
+        let parts_clone = parts.clone();
+        let body_clone = body.clone();
+        let res_sender_clone = res_tx.clone();
+        tokio::spawn(async move {
```

Comment:
the result of this future is immediately awaited on, why are we spawning a task here?

Reply:
If that's the case maybe we should drop a comment explaining it? Or consider an explicit cancellation token